tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -2671,13 +2748,116 @@ LogicalResult ConvertTFLPReluOp::matchAndRewrite(
+  //   out = select(a2, x, a1)
+  //
+  // If alpha can be constrained to 0.0 <= alpha <= 1.0, then
+  // an alternative simpler lowering could be implemented with:
+  //
+  //   max(mul(x, alapha), x)
+  //
+  // But this alternative is not robust unless alpha meets those constraints.
+
+  FloatAttr tmpAttr = tfl_leakyrelu_op.alphaAttr();
+  // There is disagreement between the MLIR .td defaults and TF
+  // documentation on 0.2 vs 0.3, but 0.2 will be used here.
+  double alpha = 0.2;
+
+  if (tmpAttr) {
+    alpha = tmpAttr.getValueAsDouble();
+  }
+
+  Value output;
+
+  if (output_is_qtype) {
+    // op1 = rescale(input)
+    // rescaled_alpha = (alpha << alpha_shift) // Remains within int32 range
+    // op2 = mul(rescaled_input, rescaled_alpha, alpha_shift)
+    // op3 = greater_equal(op1, 0)
+    // op4 = select(op3, op1, op2)
+    // out = rescale(op4)
+    RankedTensorType rescale_type =
+        RankedTensorType::get(output_type.getShape(), rewriter.getI32Type());
+
+    UniformQuantizedType input_qtype =
+        input_type.getElementType().cast<UniformQuantizedType>();
+
+    UniformQuantizedType output_qtype =
+        output_type.getElementType().cast<UniformQuantizedType>();
+
+    double scale_alpha =
+        input_qtype.getScale() * alpha / output_qtype.getScale();
+    double scale_identity = input_qtype.getScale() / output_qtype.getScale();
+
+    Value op1_rescale_in =
+        buildRescaleToInt32(rewriter, op, tfl_leakyrelu_op.input(), 1.0,
+                            input_qtype.getZeroPoint());
+
+    Value const_zero = getTosaConstTensorSingleI32(rewriter, op, 0);
+
+    auto op2_ge = rewriter.create<tosa::GreaterEqualOp>(
+        op->getLoc(),
+        RankedTensorType::get(rescale_type.getShape(), rewriter.getI1Type()),
+        op1_rescale_in, const_zero);
+
+    Value op3_rescale_alpha_in = buildRescale(
+        rewriter, op, output_type, tfl_leakyrelu_op.input(), scale_alpha,
+        input_qtype.getZeroPoint(), output_qtype.getZeroPoint(), true);
+
+    Value op4_rescale_identity_in = buildRescale(
+        rewriter, op, output_type, tfl_leakyrelu_op.input(), scale_identity,
+        input_qtype.getZeroPoint(), output_qtype.getZeroPoint(), true);
+
+    auto op5_select = rewriter.create<tosa::SelectOp>(
+        op->getLoc(), output_type, op2_ge, op4_rescale_identity_in,
+        op3_rescale_alpha_in);
+
+    output = op5_select.getResult();
+
+  } else {
+    Value const_zero = getTosaConstTensorSingleF32(rewriter, op, 0.0);
+
+    auto op1_mul = rewriter.create<tosa::MulOp>(
+        op->getLoc(), output_type, tfl_leakyrelu_op.input(),
+        getTosaConstTensorSingleF32(rewriter, op, alpha), 0);
+
+    auto op2_ge = rewriter.create<tosa::GreaterEqualOp>(
+        op->getLoc(),
+        RankedTensorType::get(output_type.getShape(),
+                              rewriter.getIntegerType(1)),
+        tfl_leakyrelu_op.input(), const_zero);
+
+    auto op3_select = rewriter.create<tosa::SelectOp>(
```

Comment:
Nit: Prefer that each branch of the if just end with a `rewriter.replaceOpWithNewOp<...>(...)` and a `return success();` (versus ending with a shared `replaceOp` and return after the condition).

Reply:
Thanks. Will fix in next round.